File: pkg/app/match_key.go
package app

import (
	"errors"
	"strings"
)

// MatchKey is the typed form of the stringly-typed `match_key`
// identity used across the codebase. A match key is one of three
// shapes:
//
//   - `match-<YYYY-MM-DDTHH-MM-SS>`  → KindTracked     (real match)
//   - `unmatched-<filename>`         → KindUnmatched   (timestamp absent)
//   - `ambiguous-<filename>`         → KindAmbiguous   (pending resolution)
//
// The legacy form used `:` instead of `-` between the kind prefix and
// the body; a one-time migration in pkg/db.SQLStore rewrote on-disk
// rows so the wire format is now stable on the dash form.
//
// This type is OPT-IN at internal call sites today — consumers can
// keep treating match_key as a bare string for back-compat. New code
// that needs to BRANCH on the kind (e.g. "is this an ambiguous
// row?") should parse via ParseMatchKey and switch on .Kind rather
// than `strings.HasPrefix`. The eventual goal is to migrate every
// startsWith / HasPrefix site to this type — tracked in
// TECHNICAL_DEBT.md.

type MatchKeyKind int

const (
	KindInvalid MatchKeyKind = iota
	KindTracked
	KindUnmatched
	KindAmbiguous
)

type MatchKey struct {
	Kind MatchKeyKind
	// Raw is the original input string, preserved so String() round-
	// trips. Cheaper than re-formatting from Kind + Body, and lets
	// the type carry forward shape oddities (case, trailing dots)
	// rather than swallowing them.
	Raw string
	// Body is the portion of the key past the kind prefix. For
	// KindTracked this is the ISO-extended timestamp (with `-`
	// separators); for the other kinds it's the original filename.
	Body string
}

// ErrInvalidMatchKey is returned by ParseMatchKey for any input that
// doesn't carry one of the three known prefixes. The caller can
// errors.Is against this sentinel for graceful handling.
var ErrInvalidMatchKey = errors.New("invalid match key")

// ParseMatchKey returns the typed form of `s`, or ErrInvalidMatchKey
// if `s` doesn't carry one of the three known prefixes.
func ParseMatchKey(s string) (MatchKey, error) {
	switch {
	case strings.HasPrefix(s, "match-"):
		return MatchKey{Kind: KindTracked, Raw: s, Body: s[len("match-"):]}, nil
	case strings.HasPrefix(s, "unmatched-"):
		return MatchKey{Kind: KindUnmatched, Raw: s, Body: s[len("unmatched-"):]}, nil
	case strings.HasPrefix(s, "ambiguous-"):
		return MatchKey{Kind: KindAmbiguous, Raw: s, Body: s[len("ambiguous-"):]}, nil
	}
	return MatchKey{}, ErrInvalidMatchKey
}

// String returns the wire form of the key. Round-trips through
// ParseMatchKey unchanged.
func (k MatchKey) String() string { return k.Raw }

// IsAmbiguous is a thin convenience over Kind == KindAmbiguous —
// the most common branch at every existing call site. Keeps the
// .Kind enum private at the consumer.
func (k MatchKey) IsAmbiguous() bool { return k.Kind == KindAmbiguous }

// IsUnmatched mirrors IsAmbiguous for unmatched keys.
func (k MatchKey) IsUnmatched() bool { return k.Kind == KindUnmatched }

// IsTracked mirrors IsAmbiguous for normal tracked keys.
func (k MatchKey) IsTracked() bool { return k.Kind == KindTracked }

File: frontend/src/match/match-key.ts
// Typed form of the stringly-typed `match_key` identity. Mirrors
// pkg/match/match_key.go on the Go side — same kind enum, same four
// prefixes, same Filename / String helpers. The wire format is the
// authority; this module is the parser + constructors that keep
// consumers from open-coding `strings.startsWith('ambiguous-')`.
//
// `replay` is the newest kind and the only one the parser never mints: it
// is derived from the six characters of a replay code, so a coach and a
// player on different machines arrive at the same key. See the Go file for
// why that correspondence is the whole point.
//
// Adopt this at any new site that branches on the prefix. Existing
// sites migrate one PR at a time — the helpers are non-breaking.

type MatchKeyKind = 'tracked' | 'unmatched' | 'ambiguous' | 'replay'

export interface MatchKey {
  kind: MatchKeyKind
  // The original wire-format string, preserved so toString round-
  // trips. Cheaper than re-formatting from kind + body, and lets
  // shape oddities (case, trailing dots) carry forward rather than
  // get swallowed in the parse → render cycle.
  raw: string
  // The portion after the kind prefix — the timestamp for tracked
  // keys; the base64url-encoded filename for unmatched / ambiguous
  // (the Go side decodes it; the frontend treats it as opaque); the
  // canonical replay code for replay keys.
  body: string
}

export class InvalidMatchKeyError extends Error {
  constructor(public readonly input: string) {
    super(`invalid match key: ${JSON.stringify(input)}`)
    this.name = 'InvalidMatchKeyError'
  }
}

/**
 * parseMatchKey returns the typed form of `s` or throws
 * InvalidMatchKeyError when `s` doesn't carry one of the four known
 * prefixes. Throwing rather than returning a union keeps the call
 * sites compact — the unknown case is genuinely exceptional, not a
 * routine branch.
 */
export function parseMatchKey(s: string): MatchKey {
  if (s.startsWith('match-')) {
    return { kind: 'tracked', raw: s, body: s.slice('match-'.length) }
  }
  if (s.startsWith('unmatched-')) {
    return { kind: 'unmatched', raw: s, body: s.slice('unmatched-'.length) }
  }
  if (s.startsWith('ambiguous-')) {
    return { kind: 'ambiguous', raw: s, body: s.slice('ambiguous-'.length) }
  }
  if (s.startsWith('replay-')) {
    return { kind: 'replay', raw: s, body: s.slice('replay-'.length) }
  }
  throw new InvalidMatchKeyError(s)
}

/**
 * tryParseMatchKey is the throw-free variant — returns null for an
 * unknown prefix. Use this when the call site wants a single ternary
 * over the result and doesn't care about distinguishing "absent"
 * from "malformed".
 */
export function tryParseMatchKey(s: string): MatchKey | null {
  try {
    return parseMatchKey(s)
  } catch {
    return null
  }
}

export function isAmbiguousMatchKey(s: string): boolean {
  return s.startsWith('ambiguous-')
}

export function isUnmatchedMatchKey(s: string): boolean {
  return s.startsWith('unmatched-')
}

export function isTrackedMatchKey(s: string): boolean {
  return s.startsWith('match-')
}

export function isReplayMatchKey(s: string): boolean {
  return s.startsWith('replay-')
}

/**
 * Whether a note may be written about this match — the frontend twin of
 * coach.IsReviewableMatchKey.
 *
 * Deliberately NOT a widened `isTrackedMatchKey`: a replay match is not
 * tracked, it has no screenshot and no timestamp, and every existing caller
 * of `isTrackedMatchKey` means the narrow thing. Reviewability is a second
 * question that happens to have a wider answer.
 */
export function isReviewableMatchKey(s: string): boolean {
  return isTrackedMatchKey(s) || isReplayMatchKey(s)
}
